### PR TITLE
remove retro

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?
Fully removes retro codepaths from MLM as it deprecated and not supported anymore.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: tools/bert_embedding/embed.py [hunk 1]
```diff
@@ -1,18 +1,21 @@
...
+from types import SimpleNamespace
+from typing import Callable, Dict, List, Optional, Tuple, TypedDict
 import numpy as np
 import os
 import time
 import torch
+from torch.distributed import ProcessGroup
 from torch.utils.data import BatchSampler, DataLoader, SequentialSampler, Subset
 from torch.utils.data._utils.collate import default_collate
 from tqdm import tqdm
```
Comment (lmcafee-nvidia) [inline tools/bert_embedding/embed.py]: remove this import of `tqdm` ... there's a new conditional import of it below

File: tools/bert_embedding/embed.py [hunk 2]
```diff
@@ -127,6 +144,211 @@ def embed_data_loader(models, data_loader, tag):
...
+
+            return sampled_blocks
+
+        blocks.existing = sample_blocks(blocks.existing)
+        blocks.missing = sample_blocks(blocks.missing)
+
+    return blocks
+
+
+class Block(TypedDict):
```
Comment (lmcafee-nvidia) [inline tools/bert_embedding/embed.py]: nit: move this above `get_blocks()` to have the definition before the usage

File: megatron/legacy/model/transformer.py [hunk 3]
```diff
@@ -936,24 +924,7 @@ def __init__(self, config,
...
-                config=config,
-                model_type=ModelType.retro_encoder,
-                self_attn_mask_type=AttnMaskType.padding,
-                pre_process=True,
-                post_process=False,
-            )
-            self._retriever_key = 'retriever'
-        else:
-            self.retriever = None
+        self.retriever = None
```
Comment (lmcafee-nvidia) [inline megatron/legacy/model/transformer.py]: why are we keeping this? we should remove all references to retro

File: megatron/legacy/model/transformer.py [hunk 4]
```diff
@@ -991,180 +962,6 @@ def default_decoder_cross_attention(self,
...
-            norm_input = args.retro_attention_gate * norm_input + residual
-
-        # Layer norm post the decoder attention
-        norm_output = self.post_inter_attention_norm(norm_input)
-
-        return retriever_output, norm_input, norm_output
-
     def forward(self, hidden_states, attention_mask,
                 encoder_output=None, enc_dec_attn_mask=None,
                 retriever_input=None,
```
Comment (lmcafee-nvidia) [inline megatron/legacy/model/transformer.py]: remove these 3 `retriever_*` args

File: megatron/legacy/model/transformer.py [hunk 5]
```diff
@@ -1783,9 +1508,6 @@ def forward(self, hidden_states, attention_mask,
                             attention_mask,
                             **forward_kwargs)
 
-                        # First Retro decoder layer returns both hidden_states
-                        # and retriever_output. Make retriever_output available
-                        # to subsequence Retro layers.
                         if isinstance(hidden_states, tuple):
                             assert len(hidden_states) == 2
                             hidden_states, retriever_output = hidden_states
```
Comment (lmcafee-nvidia) [inline megatron/legacy/model/transformer.py]: `hidden_states` should no longer contain `retriever_output` ... and I think generally should no longer be a tuple

File: megatron/legacy/model/transformer.py [hunk 6]
```diff
@@ -1364,20 +1129,8 @@ def _get_num_layers(args, model_type, is_decoder=False):
...
-            return LayerType.retro_decoder_with_retriever \
-                if layer_number == retro_layer_numbers[0] \
-                   else LayerType.retro_decoder
-        elif model_type == ModelType.retro_encoder:
-            return LayerType.retro_encoder
-        else:
-            raise Exception("Unsupported model type, '%s'." % model_type)
-    else:
-        return default_layer_type
+def _get_layer_type(default_layer_type):
```
Comment (lmcafee-nvidia) [inline megatron/legacy/model/transformer.py]: why are we keeping a function that returns its argument?

File: megatron/legacy/model/language_model.py [hunk 7]
```diff
@@ -360,7 +360,7 @@ def __init__(
         self.decoder_attn_mask_type = decoder_attn_mask_type
         self.add_pooler = add_pooler
         self.encoder_hidden_state = None
-        self.add_retriever = args.retro_add_retriever
+        self.add_retriever = False
```
Comment (lmcafee-nvidia) [inline megatron/legacy/model/language_model.py]: this should be removed

File: docs/api-guide/models/models.md [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # models package
 
-This package contains most of the popular LLMs . Currently we have support for GPT, Bert, T5 and Retro . This is an ever growing list so keep an eye out.
+This package contains most of the popular LLMs . Currently we have support for GPT, Bert, and T5 . This is an ever growing list so keep an eye out.
```
Comment (ericharper) [inline docs/api-guide/models/models.md:3]: Not for this PR but in general @Phlip79 , this is outdated.

Comment (Phlip79) [reply docs/api-guide/models/models.md:3]: Added this as a TODO to #2619.
